RFC0014 - Arrays and varargs (#119)
Draco-lang/Language-suggestions: Specification/ControlFlow.md

There will be two syntax variants, the "block syntax":

```swift
func function_name(arg1: Type1, arg2: Type2): ReturnType {
    // Function body
}
```

And the "inline-expression" syntax (roughly equivalent to arrow-bodied methods in C#):

```swift
func function_name(arg1: Type1, arg2: Type2): ReturnType = expression;
```

The `: ReturnType` part is optional for both syntaxes.

> The parameter list can optionally contain a single vararg, such vararg must be placed at the end of the argument list. Varargs have the syntax `...args: Array<T>`.  

Reviewer: How about supporting `...args: Span<T>`?

Author: I would imagine we could support any collection for future, but for now the idea of collections is not really specified, so to make things simple we are just specifiing it for arrays now.

Reviewer 2: Since C# is also planning to support spans, we'll likely follow to keep up with API usage.

The reason the current spec is so constrained is because we have started implementing arrays in the compiler and we wanted some way to have easy array initializers, varargs being a very simple way of providing that.

Good forward thinking tho, and this was the reason I wanted the collection type to be explicit like in C#, and not like Java where it's materialized by the compiler

> The parameter list can optionally contain a single vararg, such vararg must be placed at the end of the argument list. Varargs have the syntax `...args: Array<T>`.  

Reviewer 2: ```suggestion
The parameter list can optionally contain a variadic argument list as the last parameter, which must have the `...parameterName: CollectionType` declaration syntax. 
```

There will be two syntax variants, the "block syntax":

```swift
func function_name(arg1: Type1, arg2: Type2): ReturnType {
    // Function body
}
```

And the "inline-expression" syntax (roughly equivalent to arrow-bodied methods in C#):

```swift
func function_name(arg1: Type1, arg2: Type2): ReturnType = expression;
```

The `: ReturnType` part is optional for both syntaxes.

The parameter list can optionally contain a variadic argument list as the last parameter, which must have the `...parameterName: CollectionType` declaration syntax. 
```swift
func function_name(arg1: Type1, arg2: Type2, ...args: Array<Type3>)
```

> If a collection is passed as a variadic argument, the collection won't be passed as individual elements contained in the collection, but as a single element into the collection of variadic arguments. If the given collection should be passed as the elements it contains the splat operator must be used.

Reviewer 2: ```suggestion
If a collection is passed as a variadic argument, the collection won't be passed as individual elements - like in C# -, but as a single element instead. If the given collection should be passed as the individual elements it contains, the splat operator must be used.
```

}
```

> Varargs are used as regular Array inside the function. Calling a function that contains vararg allows the user to append an arbitrary number of arguments that have the same type as the array of the varargs. 

Reviewer 2: ```suggestion
Variadic arguments can be used as regular collections inside the function - as their type specified. While currently only variadic argument lists with type `Array<T>` are supported, in the future this can be extended to more collections and even spans. Calling a function that contains a variadic argument list allows the user to append an arbitrary number of arguments that has the same type as the element type of the variadic argument collection. 
```

}
```

Variadic arguments can be used as regular collections inside the function - as their type specified. While currently only variadic argument lists with type `Array<T>` are supported, in the future this can be extended to more collections and even spans. Calling a function that contains a variadic argument list allows the user to append an arbitrary number of arguments that has the same type as the element type of the variadic argument collection. 
```swift
func foo(arg1: string, ...args: Array<int32>) 
{
    var result = 0;
    var i = 0;
    while(i < args.Length){
        result += args[i];
        i += 1;
    }
    return result;
}

func main(){
    foo("Hi", 5); // OK
    foo("Hello", 5, 10, 15, 25); // OK
    foo(5, 10, 15, 25); // ERROR: all regular arguments still must be provided

> }

Reviewer 2: We could add a nullary example here
